### ISSUE TITLE
pygame and kivmob

### ISSUE THREAD

**Issue by SonofPoseidon1982** [MichaelStott/KivMob#95]
Can you use kivmob with pygame? You see I've managed to get Unity ads working with pyjnius - it works with my Kivy app but crashes with a black screen death on pygame.

**Comment by MichaelStott** [comment]
It might be possible with some fiddling. I don't know how to go about building a pygame project for Android, but it doesn't look like it uses buildozer. As a result, you willl need to manually add the AdMob library/dependencies to the Android project and make sure the app permissions are set accordingly.  `kivmob.py` also references the Kivy logging module, which may cause some issues when building a pygame project, but that could easily be replaced or abstracted away. 

This would need more investigation, but I like the idea of having this library include pygame support/build instructions. Marking as a possible enhancement for now.

**Comment by SonofPoseidon1982** [comment]
It's ok :) Pygame now has a buildozer recipe and works fine. It was introduced this year. There are no problems with kivy logging module as I've used it in my pygame project. 

I imported unity ads with pyjnius, it worked absolutely flawlessly in Kivy. But in pygame, there were problems with display the ads.. they would load,  but then the black screen death occurred.

**Comment by Sahil-pixel** [comment]
hey..can you tell me ..how you have converted  pygame game into apk..?
can you share your buildozer.spec file ?

**Comment by Maxython** [comment]
@Sahil-pixel, buildozer is a compiler for Kivy.  Though I've never done PyGame, try this https://github.com/renpytom/rapt-pygame-example.

**Comment by Sahil-pixel** [comment]
it is only for python 2

**Comment by Maxython** [comment]
Yes, it needs python 2. At the moment there is no compiler for PyGame in python 3. There are only two solutions: 
1) Switch to kivy.  
2) Compile PyGame on python 2 (you can use linux for this).

**Comment by Sahil-pixel** [comment]
I know Kivy see my game written in Kivy .
https://play.google.com/store/apps/details?id=org.flappypybird.flappypybird

**Comment by Maxython** [comment]
Self-advertisement, classic) 
The game is cool, but did I answer your question?

**Comment by Sahil-pixel** [comment]
yes .I want to use pygame ..I don't have any idea with pygame to convert into an APK ....and there is no good tutorial ...

**Comment by SonofPoseidon1982** [comment]
Sahil-pixel, you can use pygame in buildozer. I have used it myself. There is a recipe. See for yourself; https://github.com/kivy/python-for-android/tree/master/pythonforandroid/recipes

Just include it in your list of recipes (It works with python3). This was added this year.

**Comment by SonofPoseidon1982** [comment]
p4a recipes are not limited to Kivy, even though that is an important interface. I previously used Kivy (for simple apps). but I find pygame more conducive for games (I'm having problems getting ads in which worked on Kivy though)

**Comment by SonofPoseidon1982** [comment]
I will share my buildozer spec file, so here it is:

`[app]

# (str) Title of your application
title = Money Rain

# (str) Package name
package.name = myapp

# (str) Package domain (needed for android/ios packaging)
package.domain = com.mrain

# (str) Source code where the main.py live
source.dir = .

# (list) Source files to include (let empty to include all the files)
source.include_exts = py,png,jpg,kv,atlas,dat,ogg

# (list) List of inclusions using pattern matching
#source.include_patterns = assets/*,images/*.png

# (list) Source files to exclude (let empty to not exclude anything)
#source.exclude_exts = spec

# (list) List of directory to exclude (let empty to not exclude anything)
#source.exclude_dirs = tests, bin

# (list) List of exclusions using pattern matching
#source.exclude_patterns = license,images/*/*.jpg

# (str) Application versioning (method 1)
version = 0.1

# (str) Application versioning (method 2)
# version.regex = __version__ = ['"](.*)['"]
# version.filename = %(source.dir)s/main.py

# (list) Application requirements
# comma separated e.g. requirements = sqlite3,kivy
requirements = python3,pygame,hostpython3,numpy,jnius,sdl2,sdl2_image,sdl2_mixer,sdl2_ttf,kivy

# (str) Custom source folders for requirements
# Sets custom source for any requirements with recipes
# requirements.source.kivy = ../../kivy

# (list) Garden requirements
#garden_requirements =

# (str) Presplash of the application
presplash.filename = %(source.dir)s/open.png

# (str) Icon of the application
icon.filename = %(source.dir)s/icon.png

# (str) Supported orientation (one of landscape, sensorLandscape, portrait or all)
orientation = landscape

# (list) List of service to declare
#services = NAME:ENTRYPOINT_TO_PY,NAME2:ENTRYPOINT2_TO_PY

#
# OSX Specific
#

#
# author = © Copyright Info

# change the major version of python used by the app
osx.python_version = 3

# Kivy version to use
osx.kivy_version = 1.9.1

#
# Android specific
#

# (bool) Indicate if the application should be fullscreen or not
fullscreen = 1

# (string) Presplash background color (for new android toolchain)
# Supported formats are: #RRGGBB #AARRGGBB or one of the following names:
# red, blue, green, black, white, gray, cyan, magenta, yellow, lightgray,
# darkgray, grey, lightgrey, darkgrey, aqua, fuchsia, lime, maroon, navy,
# olive, purple, silver, teal.
#android.presplash_color = #FFFFFF

# (list) Permissions
android.permissions = INTERNET,ACCESS_NETWORK_STATE

# (int) Target Android API, should be as high as possible.
android.api = 29

# (int) Minimum API your APK will support.
android.minapi = 21

# (int) Android SDK version to use
android.sdk = 24

# (str) Android NDK version to use
android.ndk = 19b

# (int) Android NDK API to use. This is the minimum API your app will support, it should usually match android.minapi.
android.ndk_api = 21

# (bool) Use --private data storage (True) or --dir public storage (False)
#android.private_storage = True

# (str) Android NDK directory (if empty, it will be automatically downloaded.)
#android.ndk_path =

# (str) Android SDK directory (if empty, it will be automatically downloaded.)
#android.sdk_path =

# (str) ANT directory (if empty, it will be automatically downloaded.)
#android.ant_path =

# (bool) If True, then skip trying to update the Android sdk
# This can be useful to avoid excess Internet downloads or save time
# when an update is due and you just want to test/build your package
# android.skip_update = False

# (bool) If True, then automatically accept SDK license
# agreements. This is intended for automation only. If set to False,
# the default, you will be shown the license when first running
# buildozer.
# android.accept_sdk_license = False

# (str) Android entry point, default is ok for Kivy-based app
#android.entrypoint = org.renpy.android.PythonActivity

# (str) Android app theme, default is ok for Kivy-based app
#android.apptheme = "@android:style/Theme.NoTitleBar"

# (list) Pattern to whitelist for the whole project
#android.whitelist =

# (str) Path to a custom whitelist file
#android.whitelist_src =

# (str) Path to a custom blacklist file
#android.blacklist_src =

# (list) List of Java .jar files to add to the libs so that pyjnius can access
# their classes. Don't add jars that you do not need, since extra jars can slow
# down the build process. Allows wildcards matching, for example:
# OUYA-ODK/libs/*.jar
#android.add_jars = foo.jar,bar.jar,path/to/more/*.jar

# (list) List of Java files to add to the android project (can be java or a
# directory containing the files)
#android.add_src =

# (list) Android AAR archives to add (currently works only with sdl2_gradle
# bootstrap)
android.add_aars = ./libs/*.aar

# (list) Gradle dependencies to add (currently works only with sdl2_gradle
# bootstrap)
#android.gradle_dependencies =

# (list) add java compile options
# this can for example be necessary when importing certain java libraries using the 'android.gradle_dependencies' option
# see https://developer.android.com/studio/write/java8-support for further information
# android.add_compile_options = "sourceCompatibility = 1.8", "targetCompatibility = 1.8"

# (list) Gradle repositories to add {can be necessary for some android.gradle_dependencies}
# please enclose in double quotes 
# e.g. android.gradle_repositories = "maven { url 'https://kotlin.bintray.com/ktor' }"
#android.add_gradle_repositories =

# (list) packaging options to add 
# see https://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.PackagingOptions.html
# can be necessary to solve conflicts in gradle_dependencies
# please enclose in double quotes 
# e.g. android.add_packaging_options = "exclude 'META-INF/common.kotlin_module'", "exclude 'META-INF/*.kotlin_module'"
#android.add_gradle_repositories =

# (list) Java classes to add as activities to the manifest.
#android.add_activities = com.example.ExampleActivity

# (str) OUYA Console category. Should be one of GAME or APP
# If you leave this blank, OUYA support will not be enabled
#android.ouya.category = GAME

# (str) Filename of OUYA Console icon. It must be a 732x412 png image.
android.ouya.icon.filename = %(source.dir)s/opening.png

# (str) XML file to include as an intent filters in <activity> tag
#android.manifest.intent_filters =

# (str) launchMode to set for the main activity
#android.manifest.launch_mode = standard

# (list) Android additional libraries to copy into libs/armeabi
#android.add_libs_armeabi = libs/android/*.so
#android.add_libs_armeabi_v7a = libs/android-v7/*.so
#android.add_libs_arm64_v8a = libs/android-v8/*.so
#android.add_libs_x86 = libs/android-x86/*.so
#android.add_libs_mips = libs/android-mips/*.so

# (bool) Indicate whether the screen should stay on
# Don't forget to add the WAKE_LOCK permission if you set this to True
#android.wakelock = False

# (list) Android application meta-data to set (key=value format)
#android.meta_data =

# (list) Android library project to add (will be added in the
# project.properties automatically.)
#android.library_references =

# (list) Android shared libraries which will be added to AndroidManifest.xml using <uses-library> tag
#android.uses_library =

# (str) Android logcat filters to use
#android.logcat_filters = *:S python:D

# (bool) Copy library instead of making a libpymodules.so
#android.copy_libs = 1

# (str) The Android arch to build for, choices: armeabi-v7a, arm64-v8a, x86, x86_64
android.arch = armeabi-v7a

# (int) overrides automatic versionCode computation (used in build.gradle)
# this is not the same as app version and should only be edited if you know what you're doing
# android.numeric_version = 1

#
# Python for android (p4a) specific
#

# (str) python-for-android fork to use, defaults to upstream (kivy)
#p4a.fork = kivy

# (str) python-for-android branch to use, defaults to master
p4a.branch = master

# (str) python-for-android git clone directory (if empty, it will be automatically cloned from github)
#p4a.source_dir =

# (str) The directory in which python-for-android should look for your own build recipes (if any)
#p4a.local_recipes =

# (str) Filename to the hook for p4a
#p4a.hook =

# (str) Bootstrap to use for android builds
# p4a.bootstrap = sdl2

# (int) port number to specify an explicit --port= p4a argument (eg for bootstrap flask)
#p4a.port =


#
# iOS specific
#

# (str) Path to a custom kivy-ios folder
#ios.kivy_ios_dir = ../kivy-ios
# Alternately, specify the URL and branch of a git checkout:
ios.kivy_ios_url = https://github.com/kivy/kivy-ios
ios.kivy_ios_branch = master

# Another platform dependency: ios-deploy
# Uncomment to use a custom checkout
#ios.ios_deploy_dir = ../ios_deploy
# Or specify URL and branch
ios.ios_deploy_url = https://github.com/phonegap/ios-deploy
ios.ios_deploy_branch = 1.7.0

# (str) Name of the certificate to use for signing the debug version
# Get a list of available identities: buildozer ios list_identities
#ios.codesign.debug = "iPhone Developer: <lastname> <firstname> (<hexstring>)"

# (str) Name of the certificate to use for signing the release version
#ios.codesign.release = %(ios.codesign.debug)s


[buildozer]

# (int) Log level (0 = error only, 1 = info, 2 = debug (with command output))
log_level = 2

# (int) Display warning if buildozer is run as root (0 = False, 1 = True)
warn_on_root = 1

# (str) Path to build artifact storage, absolute or relative to spec file
# build_dir = ./.buildozer

# (str) Path to build output (i.e. .apk, .ipa) storage
# bin_dir = ./bin

#    -----------------------------------------------------------------------------
#    List as sections
#
#    You can define all the "list" as [section:key].
#    Each line will be considered as a option to the list.
#    Let's take [app] / source.exclude_patterns.
#    Instead of doing:
#
#[app]
#source.exclude_patterns = license,data/audio/*.wav,data/images/original/*
#
#    This can be translated into:
#
#[app:source.exclude_patterns]
#license
#data/audio/*.wav
#data/images/original/*
#


#    -----------------------------------------------------------------------------
#    Profiles
#
#    You can extend section / key with a profile
#    For example, you want to deploy a demo version of your application without
#    HD content. You could first change the title to add "(demo)" in the name
#    and extend the excluded directories to remove the HD content.
#
#[app@demo]
#title = My Application (demo)
#
#[app:source.exclude_patterns@demo]
#images/hd/*
#
#    Then, invoke the command line with the "demo" profile:
#
#buildozer --profile demo android debug
`

**Comment by Maxython** [comment]
As I understand it, it was just necessary to add PyGame to the requirement, and this is actually logical.

**Comment by SonofPoseidon1982** [comment]
Yup, it works very well

**Comment by SonofPoseidon1982** [comment]
I wish I could get help with my problem.. I managed to get unity ads into my Kivy App with pyjnius, but it doesn't show with a black screen in pygame. This could be due to graphical differences. Kivy has an opengl wrapper. This is all relevant if kivmob is also to work in pygame.

**Comment by SonofPoseidon1982** [comment]
Right, I'm deleting the buildozer.spec comment as its taking too much space (I think you've got what you need)

**Comment by Sahil-pixel** [comment]
i have tried see this .but i have failed.. 
https://github.com/kivy/python-for-android/issues/2333

**Comment by Rajesh-JP** [comment]
> hey..can you tell me ..how you have converted pygame game into apk..?
> can you share your buildozer.spec file ?

 It's easy to convert a pygame program into an android app. I have made a file sharing app with pygame and socket programming that worked well. At the beginning, I also had some issues like app crashing after showing presplash screen. But after thinking too many things, I came to know that the issue is in loading images and pygame display. You want to correct something in your code. 
1. Set pygame display in FULLSCREEN mode.
 Eg: surface = pygame.display.set_mode((w,h), FULLSCREEN)
2. If your app uses any images, then follow this step. Add this string before the relative path to your image. "/data/data/(package.name)/files/app/".
Here (package.name) means the value that you given to the "package.name" in buildozer.spec file.
This step is because the additional files of you app like images, audios will moved to this path while installing the app.
Eg: If your app package name is "org.test.mypgapp", then -
img = pygame.image.load("/data/data/org.test.mypgapp/files/app/myimg.png")

**Comment by SonofPoseidon1982** [comment]
Also for smooth running you will need to convert the surface to the same pixel format as the android display, so use the function convert_alpha(). For example, 

img = pygame.image.load("/data/data/org.test.mypgapp/files/app/myimg.png").convert_alpha()

I have to say however, we seem to have gone off topic as this thread was really supposed to be about pygame and kivmob :D

**Comment by Rajesh-JP** [comment]
Yes. We gone off the topic. If any one knows 
the answer please comment here. I also want to know the how to display ads in my pygame android app , that I referenced before (File sharing app).

**Comment by SonofPoseidon1982** [comment]
I've been working on it (but can't find a way around my black screen death every time the interstitial ad tries to show). Here is a link to what I've managed so far with UnityADS: [https://stackoverflow.com/questions/64522188/why-is-my-unity-ads-integration-working-on-kivy-but-not-pygame-any-ideas](url)

**Comment by Maxython** [comment]
In short, I dug around on the Internet and found an interesting fact. The PyGame module uses the SDL cross-platform multimedia library (it can also use version 2), and kivy uses SDL2 as the logic of the game, it already uses OpenGL to display the game on the screen. In short, the black eran of death may occur due to different ways of displaying the game on the screen (but this is not accurate, maybe I misunderstood something).

**Comment by SonofPoseidon1982** [comment]
I think you're probably on right lines with what you're saying. I just find it all very unclear. My games descent into blackness remains unresolved unfortunately :D

**Comment by Maxython** [comment]
@MichaelStott can you explain the logic and action of kivmod?

**Comment by MichaelStott** [comment]
@SonofPoseidon1982 Can you provide an example project demonstrating the issue?

@Maxython On a high level, kivmob instantiates a new AdMob ad view and adds it to the activity that runs the Python application through pyjnius for banner ads specifically. For interstitial and rewarded video ads, I believe AdMob handles the switching of views in that case. KivMob basically creates and manages the interactions between the interstitial object associated with PythonActivity.

**Comment by SonofPoseidon1982** [comment]
@MichaelStott  I have quickly made this project to demonstrate the problem.
[https://github.com/SonofPoseidon1982/Py_1982](url)

If you have any questions, I will try my best to answer (or anything I need to add or clarify)

**Comment by Rajesh-JP** [comment]
Please explain how the KIVMOB works in low level.

**Comment by Rajesh-JP** [comment]
Explain what it do actually

**Comment by SonofPoseidon1982** [comment]
Does anyone have any theory that might explain why I'm getting a black screen death in my pygame (taking into account that it is work smoothly with kivy)? At the moment if feelings like I'm groping around in the dark.

**Comment by Maxython** [comment]
I think I need to go to kivmod, I need to make another class that works not with kivy but with pygame.

**Comment by Maxython** [comment]
it seems to me in kivmod, you need to make another class that does not work with kivy but with pygame.

**Comment by Maxython** [comment]
sorry, working with a translator.

**Comment by SonofPoseidon1982** [comment]
Guys, I think I may have found the answer. It is to do with with FPS (Frames per second). While the interstitial ad is called during the ostensible pause (where it is not blitting graphics to the screen i.e when g_sleep==True). The frames are still refreshing at 30 FPS, and this does not stop when the ads are called.  To test this theory, I added the following line straight after 'unity_ads.show_ad(inter_id)': 
```

                unity_ads.show_ad(inter_id) 
                pygame.time.delay(40000)
```
The ad showed, but I need to find a proper solution so it can be properly integrated.

**Comment by SonofPoseidon1982** [comment]
Something like this works but its clumsy so far:

             unity_ads.show_ad(inter_id)
             pygame.time.delay(<duration of ad>)
             self.g_pause=False  # allow blitting every fps
             py_display =pygame.display.set_mode((display_width, display_height),pygame.FULLSCREEN)

This is admittedly a bit clumsy (I'm still working on it), but what it means is that ultimately kivmob should also work with pygame without any real need for revision. The methods used in kivmob are pretty much the same as mine. All you need to do is find a way to pause the updating FPS until the ad has finished showing. As for banner ads, I have no idea, that's a whole new can of worms :D

**Comment by SonofPoseidon1982** [comment]
OK, so I have found an improved solution (although it is still a little suspicious and clumsy).
 
In UnityAdsListener interface (in kivunity.py) I added 'ad_shown=True' for when the ad is shown:

    @java_method('(Ljava/lang/String;Lcom/unity3d/ads/UnityAds$FinishState;)V')
      def onUnityAdsFinish(self,inter_id,finish_state):
        #pass
        global ad_shown
        logging.warning("\n\n ADS are FINISHED! "+str(finish_state) + " \n\n")
        #appl_id = u_h.app_id2                                                     
        ad_shown = True 



I added the following function to the Unity_handler class which registers when the ad has finished (making sure to set ad_shown=False) before the ad is shown:

            def check_ad_status(self):
                return ad_shown

In main.py, I added the following:

 
           tmp_sleep=False
           unity_ads.show_ad(inter_id)
           while tmp_sleep==False:
               tmp_sleep = unity_ads.check_ad_status()
               time.sleep(1)
           self.g_sleep=False #for resuming the inner game loop
           py_display =pygame.display.set_mode((display_width, display_height),pygame.FULLSCREEN)     
           print("\n\n Pygame window back into focus \n\n ")   # Bizarrely this is sufficient to bring the pygame window back into focus


Well, my game has managed to load interstitial ads 10+ times repeatedly without any problems (no crashes). 
Although it works, my solution is not exactly elegant.

**Comment by MichaelStott** [comment]
Hey, as long as it works :)

This currently isn't exposed through KivMob for interstitials, but, you might be able to add this logic to an onShow and onClose callback for the interstitial object. That would at least remove the need to have the whole "sleep then check" loop. Callbacks for rewarded video ads are currently exposed though.

**Comment by SonofPoseidon1982** [comment]
Well this might suffice for now, while I try to find a less whacky solution :D
